torii/src/lib.rs

```diff
@@ -85,177 +94,182 @@ impl Torii {
+                    move || routing::handle_get_configuration(kiso)
                 }),
-        );
+            );
 
         #[cfg(feature = "telemetry")]
-        let get_router = get_router
-            .or(warp::path(uri::STATUS)
-                .and(add_state!(self.metrics_reporter.clone()))
-                .and(warp::header::optional(warp::http::header::ACCEPT.as_str()))
-                .and(warp::path::tail())
-                .and_then(
-                    |metrics_reporter, accept: Option<String>, tail| async move {
-                        Ok::<_, Infallible>(crate::utils::WarpResult(routing::handle_status(
+        let router = router
+            .route(
+                &format!("{}/*tail", uri::STATUS),
+                get({
+                    let metrics_reporter = self.metrics_reporter.clone();
+                    move |headers: axum::http::header::HeaderMap, axum::extract::Path(tail): axum::extract::Path<String>| {
+                        let accept = headers.get(axum::http::header::ACCEPT);
+                        core::future::ready(routing::handle_status(
                             &metrics_reporter,
-                            accept.as_ref(),
-                            &tail,
-                        )))
-                    },
-                ))
-            .or(warp::path(uri::METRICS)
-                .and(add_state!(self.metrics_reporter))
-                .and_then(|metrics_reporter| async move {
-                    Ok::<_, Infallible>(crate::utils::WarpResult(routing::handle_metrics(
-                        &metrics_reporter,
-                    )))
-                }))
-            .or(warp::path(uri::API_VERSION)
-                .and(add_state!(self.state.clone()))
-                .and_then(|state| async {
-                    Ok::<_, Infallible>(routing::handle_version(state).await)
-                }));
+                            accept,
+                            Some(&tail),
+                        ))
+                    }
+                }),
+            )
+            .route(
+                uri::STATUS,
+                get({
+                    let metrics_reporter = self.metrics_reporter.clone();
+                    move |headers: axum::http::header::HeaderMap| {
+                        let accept = headers.get(axum::http::header::ACCEPT);
+                        core::future::ready(routing::handle_status(&metrics_reporter, accept, None))
+                    }
+                }),
+            )
+            .route(
+                uri::METRICS,
+                get({
+                    let metrics_reporter = self.metrics_reporter.clone();
+                    move || core::future::ready(routing::handle_metrics(&metrics_reporter))
+                }),
+            )
+            .route(
+                uri::API_VERSION,
+                get({
+                    let state = self.state.clone();
+                    move || routing::handle_version(state)
```

Comment:
Hm, so instead of using [`State`](https://docs.rs/axum/latest/axum/extract/struct.State.html) extractor you just keep the required state bundled with the handler? Don't really see a problem with that, just never considered that you can pass closures to routes instead of bare functions.

Reply:
Yeah, it's described as one [method](https://docs.rs/axum/latest/axum/index.html#using-closure-captures) to share state with handler.